src/upload.rs

```diff
@@ -0,0 +1,264 @@
+    let enc = GzEncoder::new(tar_gz, Compression::default());
+    let mut tar = tar::Builder::new(enc);
+
+    // Add files to the tar archive
+    tar.append_dir_all(".", &project_dir)
+        .map_err(|_| UploadError::CopyFiles)?;
+
+    // Finish writing the tar archive
+    tar.finish().map_err(|_| UploadError::CopyFiles)?;
+
+    // Make sure the GzEncoder finishes and flushes all data
+    let enc = tar.into_inner().map_err(|_| UploadError::CopyFiles)?;
+    enc.finish().map_err(|_| UploadError::CopyFiles)?;
+
+    // Store the tarball in IPFS.
+    let tarball_ipfs_hash = pinata_client
+        .upload_file_to_ipfs(&final_tarball_path)
+        .await?;
+
+    fn find_file_in_dir_by_suffix(dir: &Path, suffix: &str) -> Option<PathBuf> {
+        let dir = fs::read_dir(dir).ok()?;
+        dir.flatten()
+            .filter_map(|entry| {
+                let path = entry.path();
+                if path.is_file() {
+                    if let Some(file_name) = path.file_name() {
+                        if let Some(file_name_str) = file_name.to_str() {
+                            // Check if the path is a file and ends with the suffix
+                            if file_name_str.ends_with(suffix) {
+                                return Some(path); // Return the first found file
+                            }
+                        }
+                    }
+                }
+                None
+            })
+            .next()
+    }
+
+    // Store the ABI in IPFS.
+    let abi_ipfs_hash = match find_file_in_dir_by_suffix(&release_dir, "-abi.json") {
+        Some(abi_path) => Some(pinata_client.upload_file_to_ipfs(&abi_path).await?),
+        None => None,
+    };
+
+    // Generate the bytecode identifier and store in the database along with the ABI hash.
+    let bytecode_identifier = match find_file_in_dir_by_suffix(&release_dir, ".bin") {
+        Some(bin_path) => Some(
+            get_bytecode_id(&bin_path).map_err(|err| UploadError::BytecodeId(err.to_string()))?,
+        ),
+        None => None,
+    };
+
+    let upload = NewUpload {
+        id: *upload_id,
+        source_code_ipfs_hash: tarball_ipfs_hash,
+        forc_version,
+        abi_ipfs_hash,
+        bytecode_identifier,
+    };
+
+    Ok(upload)
+}
+
+/// Installs the given version of forc at the specific root path using cargo-binstall.
+pub fn install_forc_at_path(forc_version: &str, forc_path: &Path) -> Result<(), UploadError> {
+    let os = match std::env::consts::OS {
+        "linux" => "linux",
+        "macos" => "darwin",
+        _ => return Err(UploadError::UnsupportedOs(std::env::consts::OS.to_string())),
+    };
+    let arch = match std::env::consts::ARCH {
+        "x86_64" => "amd64",
+        "aarch64" => "arm64",
+        _ => {
+            return Err(UploadError::UnsupportedArch(
+                std::env::consts::ARCH.to_string(),
+            ))
+        }
+    };
+
```

Comment:
Weird that the formatting check isn't picking this up. I tried locally to `cargo fmt` and nothing changed too. Not sure why `.arg()` doesn't need leading whitespace.

Ignore this as a requested fix I guess since fixing this might actually break the format check Action 🙃

Reply:
I noticed that too, that's how `cargo fmt` formats it.